### PR TITLE
Real time resources for agencies

### PR DESCRIPTION
**Pull request by DylanVanAssche**
This PR adds the `:agency/events` resource to the Linked Connections Server.
If a client connects to the resource, the server will check if the client is:

- **A SSE client:** add it to the listeners and pushes all the missing information since the last time the client synced with server. By watching the files directory of the events storage (generated by the EventsManager), the clients receive pushes from the server.
- **A HTTP client:** serves the events page related to the last time the client synced with the server. The client can follow the `hydra:next` properties to get the next page of events to get in sync with the server. The `lastSyncTime` is used to lookup the required page and add the `hydra` metadata to it.

Besides this resource, this PR also adds the possibility to cancel a complete trip based on its `tripURI`. If you supply a `startingConnection` the trip can be cancelled from that connection until the end. If you don't supply this argument to `cancelTrip(tripURI, agency, dataset, startingConnection=null)`, the whole trip will be cancelled from the beginning until the end.

*TO DO list finished:*
- [x] Binary search to lookup the events pages.
- [x] Ignore big sync requests, the client can asks to sync with the server with a time range of almost infinity. This might lock up the server when too many clients are doing this.
- [x] Fix the skeleton of SSE, we don't need any `hydra` metadata in there.

### CODE REVIEW

**Comment by DylanVanAssche** [comment]
Ready for review! @julianrojas87

File: lib/manager/dataset_manager.js [hunk 1]
```diff
@@ -12,6 +12,9 @@ const { Connections, Connections2JSONLD } = require('gtfs2lc');
 const jsonldstream = require('jsonld-stream');
 const pageWriterStream = require('./pageWriterStream.js');
 const { GtfsIndex, Gtfsrt2LC } = require('gtfsrt2lc');
+const patriciaTree = require('rdf-patricia-tree');
```
Comment (julianrojas87) [inline lib/manager/dataset_manager.js]: This dependency is not used here.

File: lib/manager/dataset_manager.js [hunk 2]
```diff
@@ -436,6 +461,212 @@ class DatasetManager {
...
+     * The starting connection is the first connection from where the vehicle is cancelled.
+     * If the vehicle is completely cancelled, you can omit this parameter..
+     *
+     * Example: `await this.cancelTrip('http://irail.be/vehicle/IC106/20190301', 'sncb');`
+     * @param tripURI
+     * @param companyName
+     * @param startingConnection = null
+     * @return `true` if successful
+     */
+    async cancelTrip(tripURI, companyName, startingConnection = null) {
```
Comment (julianrojas87) [inline lib/manager/dataset_manager.js]: If I understand correctly, in this method you will loop over the static pages changing the connections to `CancelledConnections`s of a certain trip. I would try to follow the same approach used for the real-time updates, where these changes are stored in separated files, keeping the static files intact and facilitating the memento process.

**Comment by julianrojas87** [comment]
Other than that I don't have any more remarks. Good job!

File: lib/manager/dataset_manager.js (hunk 2 above)
Comment (DylanVanAssche) [reply lib/manager/dataset_manager.js]: Yeah that's even better! I will look into the RT files and see how I can accomplish this.

File: lib/manager/dataset_manager.js (hunk 1 above)
Comment (DylanVanAssche) [reply lib/manager/dataset_manager.js]: This was meant for the vehicle tree, but vehicles are a `nice-to-have` in my thesis at the moment. I might add it later if my MVP experiments are complete. I will remove it for now and add the tree in another PR later.

File: lib/manager/dataset_manager.js (hunk 2 above)
Comment (DylanVanAssche) [reply lib/manager/dataset_manager.js]: Should be fixed now!

**Comment by DylanVanAssche** [comment]
@julianrojas87 I think I have fixed your last remarks, review?

**Comment by julianrojas87** [comment]
Given the last updates made to the code base, I will manually extract the contributions made on this PR and adapt them accordingly

**Comment by DylanVanAssche** [comment]
> Given the last updates made to the code base, I will manually extract the contributions made on this PR and adapt them accordingly

Fine for me!

**Comment by pietercolpaert** [comment]
Small update on this one: we are thinking more thoroughly about the hypermedia design, and including a possibility for a new Linked Data Event Streams resource on top of the existing collections

**Comment by DylanVanAssche** [comment]
@pietercolpaert Linked Data Event Streams would make a nice fit here I think :+1:

**Comment by pietercolpaert** [comment]
@DylanVanAssche While this pull request and the work behind your master thesis was instrumental to get to the LC2.0 specification, I think we will close this pull request in favour of https://github.com/linkedconnections/linked-connections-server/pull/26, in which we implement a full-history polling interface for the connections with hypermedia descriptions that follow the LDES/TREE specifications. Would you be available to review the code they have so far?

**Comment by DylanVanAssche** [comment]
> @DylanVanAssche While this pull request and the work behind your master thesis was instrumental to get to the LC2.0 specification, I think we will close this pull request in favour of #26, in which we implement a full-history polling interface for the connections with hypermedia descriptions that follow the LDES/TREE specifications. Would you be available to review the code they have so far?

Sure! I think you mean #79 ? right?

**Comment by pietercolpaert** [comment]
Yes! Edited ↑

**Pull request closed without merging**